File: src/bin/0238-product-of-array-except-self.cpp
// @leetup=info id=238 lang=cpp slug=product-of-array-except-self

#include <iostream>

// @leetup=code

#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> productExceptSelf(vector<int>& nums) {
    int product = 1;
    vector<int> prefix_product{1};
    for (int& n : nums) {
      product *= n;
      prefix_product.push_back(product);
    }

    product = 1;
    vector<int> postfix_product{1};
    for (auto it = nums.rbegin(); it != nums.rend(); ++it) {
      product *= *it;
      postfix_product.push_back(product);
    }

    vector<int> res;
    for (int i = 0; i < nums.size(); ++i) {
      res.push_back(prefix_product[i] * postfix_product[nums.size() - 1 - i]);
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0238-product-of-array-except-self.py
# @leetup=info id=238 lang=python3 slug=product-of-array-except-self



# @leetup=code
class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        N = len(nums)
        prefix_product = [1]
        product = 1
        for n in nums:
            product *= n
            prefix_product.append(product)

        postfix_product = [1]
        product = 1
        for n in reversed(nums):
            product *= n
            postfix_product.append(product)

        return [prefix_product[i] * postfix_product[N - 1 - i] for i in range(N)]


# @leetup=code
